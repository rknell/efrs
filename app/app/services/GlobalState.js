import { observable, observe, spy } from 'mobx'
import { Alert, AsyncStorage, Linking } from 'react-native'
import PushNotifications from './PushNotifications'
import Translate from '../services/Translate'

class GlobalState {
  // apiPath = 'http://192.168.1.8:1676/api'
  apiPath = 'https://www.efrs.io/api'
  // apiPath = 'http://localhost:1676/api'

  @observable model = {
    userToken: null,
    user: null,
    userPermissions: {
      permissions: {},
      groups: []
    },
    canPop: false,
    forms: [],
    cachedCalls: {
      get: {},
      post: []
    },
    isBusy: 0,
    pushToken: null
  }

  current = () => {
    return {
      user: this.model.user ? this.model.user._id : 'no user',
      group: this.model.userPermissions.groups.join(',')
    }
  }

  navigation = {
    resetTo: () => {
      console.log('RESETTING NAVIGATION')
    },
    push: () => {
      console.log('PUSHING NAVIGATION')
    }
  }

  saveSetting (key, value) {
    return AsyncStorage.setItem(`@EFRS.AppSettings:${key}`, value)
  }

  getSetting (key) {
    return AsyncStorage.getItem(`@EFRS.AppSettings:${key}`)
  }

  constructor () {
    spy(event => {
      if (event.type === 'action' || event.type === 'update') {
        // console.log(event);
      }
    })

    this.getSetting('formsData').then(data => {
      if (data) this.model.forms = JSON.parse(data)
    })

    const PNObj = new PushNotifications(this._setPushToken)
    this.pushNotifications = PNObj.messaging

    observe(this.model, 'user', () => {
      try {
        if (this.model.user && this.bugsnag) {
          this.bugsnag.setUser(this.model.user._id, `${this.model.user.firstName} ${this.model.user.lastName}`, this.model.user.email)
        }
      } catch (e) {
        console.error('could not set user for bugsnag', e)
      }
    })
  }

  _setPushToken = (token) => {
    this.model.pushToken = token
  }

  openUrl (url) {
    Linking.canOpenURL(url).then(supported => {
      if (!supported) {
        Alert.alert(Translate.do('Error'), Translate.do('Can\'t open this url on this device. Try installing an app that can handle this link type and try again.'))
      } else {
        return Linking.openURL(url)
      }
    }).catch(err => console.error('An error occurred', err))
  }
}

export default new GlobalState()
