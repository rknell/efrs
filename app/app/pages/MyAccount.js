import React, { Component } from 'react'
import { Text, View } from 'react-native'
import { observer } from 'mobx-react/native'
import { observable } from 'mobx'
import NavBar from '../components/NavBar'
import GlobalStyle from '../services/GlobalStyle'
import UserDetailView from '../components/UserDetailView'
import GlobalState from '../services/GlobalState'
import Dashboard from './Dashboard'
import IconButton from '../components/IconButton'
import api from '../api'
import Translate from '../services/Translate'

@observer
class MyAccount extends Component {
  @observable state = {
    isUserValid: false
  }

  constructor (props) {
    super(props)
    this.onSave = this.onSave.bind(this)
  }

  onSave () {
    GlobalState.model.isBusy++
    GlobalState.model.user.isTempPassword = false
    GlobalState.model.user.isSetup = true
    api.user.update()
      .then(() => {
        if (GlobalState.model.canPop) GlobalState.navigation.pop()
        else GlobalState.navigation.resetTo({ view: <Dashboard /> })
      })
      .finally(() => {
        GlobalState.model.isBusy--
      })
  }

  render () {
    const user = GlobalState.model.user
    return (
      <View style={{ flex: 1, backgroundColor: GlobalStyle.style.neutralColour }}>
        <NavBar title='My Account'
          isMenuDisabled={!GlobalState.model.user.isSetup || GlobalState.model.user.isTempPassword} />

        {!user.isSetup &&
        <Text style={{ alignSelf: 'center', textAlign: 'center', padding: 20 }}>
          {Translate.do(`Please confirm your details, and setup your password before you continue`)}
        </Text>
        }

        <UserDetailView isPersonal user={GlobalState.model.user}
          onValidationChanged={value => {
            this.state.isUserValid = value
          }} />

        <IconButton translate name='save' label='Save & Continue' onPress={this.onSave}
          isDisabled={!this.state.isUserValid} />

      </View>
    )
  }
}

export default MyAccount
