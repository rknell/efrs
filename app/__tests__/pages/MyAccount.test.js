import React from 'react'
import MyAccount from '../../app/pages/MyAccount'
import { shallow } from 'enzyme'
import GlobalState from '../../app/services/GlobalState'
import fixtures from '../../__fixtures__'
import toJSON from 'enzyme-to-json'

it('renders correctly', () => {
  GlobalState.model.user = fixtures.users.newUser
  const wrapper = shallow(<MyAccount />)
  expect(toJSON(wrapper)).toMatchSnapshot()
})
