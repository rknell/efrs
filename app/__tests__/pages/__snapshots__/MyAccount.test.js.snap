// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <NavBar
    isMenuDisabled={true}
    title="My Account"
  />
  <Text
    style={
      Object {
        "alignSelf": "center",
        "padding": 20,
        "textAlign": "center",
      }
    }
  >
    Please confirm your details, and setup your password before you continue
  </Text>
  <UserDetailView
    isPersonal={true}
    onValidationChanged={[Function]}
    user={
      Object {
        "__v": 0,
        "_id": "5a36fa6aac023a4b3bae2f22",
        "company": "5a36fa6aac023a4b3bae2f20",
        "createdAt": "2017-12-17T23:14:50.373Z",
        "domain": "example.com.au",
        "email": "user12@example.com",
        "firstName": "Admin",
        "isSetup": false,
        "lastName": "User",
        "password": "$2a$08$gtSMlrmdhH0zOeeYJzPh.egq2I65fD9WFV5PypJfNWqikrZOzFQJq",
        "pushTokens": Array [],
        "relationships": Array [],
        "updatedAt": "2017-12-17T23:14:50.390Z",
        "username": "user12@example.com",
      }
    }
  />
  <IconButton
    isDisabled={true}
    label="Save & Continue"
    name="save"
    onPress={[Function]}
    translate={true}
  />
</View>
`;
